test/dynamo/test_dynamo.py

```diff
@@ -102,6 +102,84 @@ def test_resnet18(self):
         met.metric_data('RunCachedGraphOutputData')[0], sample_count)
 
 
+class DynamoCpuFallbackTest(unittest.TestCase):
+
+  @classmethod
+  def setUpClass(self):
+    test_utils._set_rng_seed(42)
+
+  def test_operator_fallback(self):
+
+    def fn_fallback(t):
+      # As of 05/18/2023, torch.median is not lowered by PyTorch/XLA
+      return torch.median(t)
+
+    torch._dynamo.reset()
+    met.clear_counters()
+    device = xm.xla_device()
+
+    # Initial tracing
+    dynamo_fn = torch.compile(fn_fallback, backend="torchxla_trace_once")
+    t = torch.randn(5)
+    t_xla = t.to(device)
+    cpu_res = fn_fallback(t)
+    xla_dynamo_res = dynamo_fn(t_xla)
+    self.assertTrue(torch.allclose(cpu_res, xla_dynamo_res.cpu()))
+    self.assertEqual(met.metric_data('CompileTime')[0], 2)
+    self.assertEqual(met.metric_data('ExecuteTime')[0], 2)
+
+    # Second tracing
+    met.clear_counters()
+    xla_dynamo_res_2 = dynamo_fn(t_xla)
+    self.assertTrue(torch.allclose(cpu_res, xla_dynamo_res_2.cpu()))
+    self.assertEqual(met.metric_data('CompileTime')[0], 2)
+    self.assertEqual(met.metric_data('ExecuteTime')[0], 2)
+
+    # Verify that dynamo can handle different inputs
+    xla_dynamo_res_3 = dynamo_fn(t_xla * 3)
+    cpu_res_3 = fn_fallback(t * 3)
+    self.assertTrue(torch.allclose(cpu_res_3, xla_dynamo_res_3.cpu()))
+    self.assertEqual(met.metric_data('CompileTime')[0], 3)
+    self.assertEqual(met.metric_data('ExecuteTime')[0], 3)
```

Comment:
hmm why is it 3 here instead of 4? Wouldn't `dynamo_fn` fallback and execute 2 execution?

Reply:
oh ok, I think I understand why `t_xla * 3` would trigger a separate Compilation. `t_xla * 3` is actually a pending execution and we will call `mark_step` to materialize the input. If that's the case I don't understand why `ExecuteTime` is 3 then, it should be 5 I guess?

Reply:
So here's how the IR dump looks like when I run `DynamoCpuFallbackTest.test_operator_fallback` locally: https://gist.github.com/wonjoolee95/1426859ef0a9203dca71ad455e4badc8. This is also what I'm trying to figure out 😢 

Another odd thing is that when I run the fallback tests individuals, as such: `python test/dynamo/test_dynamo.py DynamoCpuFallbackTest.test_operator_fallback` and `python test/dynamo/test_dynamo.py DynamoCpuFallbackTest.test_fallback_multiple_submodules`. However, when I run them in a single run by running `python test/dynamo/test_dynamo.py`, they fail due to metric assertions (same as the failure in the CI). This makes me think there are possibly some pending IRs somewhere, but I tried to manually invoke `torch_xla._XLAC._clear_pending_irs(str(xm.xla_device()))` at the end of each unit test but still seeing the error.

Reply:
Hmmm... but I also met similar problem when adding `clear_all()`. The behavior of a single test and a set of tests are different.

Reply:
Ok, I found the issue. Since we use the same inputs for the two test cases, the second test case does not need to recompile for the CPU-fallbacked op. So if I change the inputs (say one test is `rand(5)` and the other is `rand(6)`) then the metric checking would be correct.

Reply:
Just updated the unit tests as well and now it seems like the CI is green. 

And for the original question regarding the `ExecuteTime`, @JackCaoG, this is my understanding after looking into it more -- on the 2nd tracing, we can see that both the `CompileTime` and `ExecuteTime` remain the same as the 1st tracing because the graph with the fallback op is already captured and cached. And since the `fn_fallback` function does not contain any XLA ops, the `ExecuteTime` doesn't increase. Now, for the third tracing with `x_xla * 3`, the `CompileTime` and `ExecuteTime` both increases by 1 to compute the `x_xla * 3`. But again, since the graph with the fallback op is already captured and cached and the `fn_fallback` does not contain any XLA ops, it does not need increase the `Compile` and `Execute` counters.

Reply:
@wonjoolee95 maybe we should add the explanation as comments?

Reply:
Sounds good, will do.

Reply:
I am still a bit confuse about
```
on the 2nd tracing, we can see that both the CompileTime and ExecuteTime remain the same as the 1st tracing because the graph with the fallback op is already captured
```

`ExecuteTime` should increase when a dynamo execution happened. It should increase both during the first dynamo run and subsequent execution.

Reply:
1. Let's call `met.clear_all()` instead of `met.clear_counters`, clear counter won't reset metrics.

I would expect that every call to `dynamo_fn` will trigger at least 2 execution, since there is a fallback op in the middle and XLA needs to execute the graph before and after the fallback op.

Reply:
Also trying completely understand why these metrics are having such numbers, but here is what I understand -- please let me know if there is anything that doesn't sound right.

> ExecuteTime should increase when a dynamo execution happened. It should increase both during the first dynamo run and subsequent execution.

ExecuteTime increases only when there is an XLA execution. However, in this example, there is only one single `aten::median` op that is executed by CPU, so the `ExecuteTime` doesn't increase on the 2nd tracing. 

> I would expect that every call to dynamo_fn will trigger at least 2 execution, since there is a fallback op in the middle and XLA needs to execute the graph before and after the fallback op.

On the 2nd and 3rd tracing, the `torchxla.py` already has a `compiled_graph` from the 1st tracing that looks like:
```
class GraphModule(torch.nn.Module):
    def forward(self, L_t_ : torch.Tensor):
        l_t_ = L_t_
        
        # File: wonjoo_2.py:28, code: return torch.median(t)
        median = torch.median(l_t_);  l_t_ = None
        return (median,)
```
Now, since the only op in this is `median` that is executed by CPU, executing this graph with `compiled_graph(*args)` does not increase `ExecuteTime`. In the 3rd tracing with `t_xla * 3`, the only graph that XLA needs to execute before the `aten::median` op is the `x_xla * 3`, hence increasing `ExecuteTime` only by 1.

This is what I think is happening.. let me know if that makes sense, @JackCaoG . And @seanlatias, also let me know if this aligns with your understanding, just to make sure we're on the same page.

Reply:
Mostly agree with @wonjoolee95. Following is my explanation for each trace.
* The first trace includes two parts: checking CPU fallback OP for compilation and run the fallback OP with the compiled graph. Both parts involve compiling and executing `torch.median()`. Thus, the initial `CompileTime` and `ExecuteTime` are 2.
* The second trace does not introduce any changes: same input & same module. Thus, nothing needs to be compiled and executed again. This is handled by the cached computation in XLA.
* The third trace does not change the input module. Thus, the module does not need to be recompiled. However, it changes the input for the fallback OP. Thus, the fallback OP needs to be recompiled and re-exectued (remember that we said the fallback OP goes through the lazy tensor execution instead of being executed by PyTorch directly). That's why we see an increase in both `CompileTime` and `ExecuteTime`.

Reply:
So the only difference in opinion between me and @wonjoolee95 is that for the third trace, I think the `CompileTime` and `ExecuteTime` increase due to `media(x_xla*3)` not `x_xla*3`. If we don't have any fallback OP, there won't be any new `CompileTime` increase.

Reply:
Make sense, thanks.